File: lib/services/eld_service.dart
import 'package:flutter/foundation.dart';
import 'package:semitrack_mobile/core/api_client.dart';

enum EldProvider { samsara, motive }

extension on EldProvider {
  String get apiName => name.toUpperCase();
}

class EldConnection {
  const EldConnection({
    required this.provider,
    required this.status,
    required this.scopes,
    this.lastSyncedAt,
    this.lastErrorMessage,
  });

  final EldProvider provider;
  final String status;
  final List<String> scopes;
  final DateTime? lastSyncedAt;
  final String? lastErrorMessage;

  factory EldConnection.fromJson(Map<String, dynamic> json) => EldConnection(
        provider: EldProvider.values.firstWhere(
          (provider) => provider.apiName == json['provider'],
        ),
        status: json['status']?.toString() ?? 'DISCONNECTED',
        scopes: (json['scopes'] as List? ?? const [])
            .map((value) => value.toString())
            .toList(growable: false),
        lastSyncedAt: DateTime.tryParse(json['lastSyncedAt']?.toString() ?? ''),
        lastErrorMessage: json['lastErrorMessage']?.toString(),
      );
}

class EldService extends ChangeNotifier {
  EldService(this.api);
  final ApiClient api;

  bool loading = false;
  String? error;
  List<EldConnection> connections = const [];
  Map<String, dynamic>? lastSnapshot;

  Future<void> load() async {
    loading = true;
    error = null;
    notifyListeners();
    try {
      final response = await api.getJson('/eld/connections');
      connections = (response['items'] as List? ?? const [])
          .map((value) => EldConnection.fromJson(value as Map<String, dynamic>))
          .toList(growable: false);
    } catch (value) {
      error = value.toString();
    } finally {
      loading = false;
      notifyListeners();
    }
  }

  Future<Uri> beginConnect(EldProvider provider) async {
    final response = await api.postJson('/eld/${provider.apiName}/connect', {});
    return Uri.parse(response['authorizeUrl'].toString());
  }

  Future<void> sync(EldProvider provider) async {
    loading = true;
    error = null;
    notifyListeners();
    try {
      lastSnapshot = await api.postJson('/eld/${provider.apiName}/sync', {});
      await load();
    } catch (value) {
      error = value.toString();
      loading = false;
      notifyListeners();
    }
  }

  Future<void> disconnect(EldProvider provider) async {
    await api.delete('/eld/${provider.apiName}');
    await load();
  }
}

File: apps/mobile-react-native/src/features/eld/EldService.ts
import { z } from 'zod';
import type { ApiClient } from '../../services/api/ApiClient';
export const eldProviders = ['SAMSARA', 'MOTIVE'] as const;
export type EldProvider = (typeof eldProviders)[number];
const connectionSchema = z.object({
  provider: z.enum(eldProviders),
  status: z.string(),
  lastSyncedAt: z.string().nullable(),
  lastErrorCode: z.string().nullable().optional(),
});
export type EldConnection = z.infer<typeof connectionSchema>;
export function authorizedEldUrl(provider: EldProvider, raw: string) {
  const url = new URL(raw);
  const expected =
    provider === 'SAMSARA'
      ? 'https://api.samsara.com/oauth2/authorize'
      : 'https://gomotive.com/oauth/authorize';
  if (
    url.origin + url.pathname !== expected ||
    url.username ||
    url.password ||
    url.hash ||
    url.searchParams.get('response_type') !== 'code' ||
    !url.searchParams.get('state') ||
    !url.searchParams.get('client_id')
  )
    throw new Error('ELD authorization address is invalid.');
  return url.toString();
}
export class EldService {
  constructor(private api: ApiClient) {}
  async connections(signal?: AbortSignal) {
    return z
      .object({ items: z.array(connectionSchema) })
      .parse(
        await this.api.request('GET', '/eld/connections', undefined, signal),
      ).items;
  }
  async connect(provider: EldProvider) {
    const data = z
      .object({ authorizeUrl: z.string() })
      .parse(
        await this.api.request('POST', '/eld/' + provider + '/connect', {}),
      );
    return authorizedEldUrl(provider, data.authorizeUrl);
  }
  async sync(provider: EldProvider) {
    await this.api.request('POST', '/eld/' + provider + '/sync', {});
  }
  async disconnect(provider: EldProvider) {
    await this.api.request('DELETE', '/eld/' + provider);
  }
  async hos(signal?: AbortSignal) {
    // Current API deliberately withholds clocks until signed-in driver mapping is proven.
    const data = z
      .object({
        status: z.string(),
        reason: z.string().optional(),
        certifiedEld: z.boolean(),
      })
      .parse(
        await this.api.request('GET', '/eld/hos/current', undefined, signal),
      );
    return {
      status: 'UNKNOWN' as const,
      reason: data.reason ?? 'DRIVER_MAPPING_REQUIRED',
    };
  }
}
